arroyo-worker/src/connectors/websocket.rs

```diff
@@ -0,0 +1,237 @@
+                    }
+                    StopMode::Immediate => {
+                        return Some(SourceFinishType::Immediate);
+                    }
+                }
+            }
+        }
+        None
+    }
+
+    async fn run(&mut self, ctx: &mut Context<(), T>) -> SourceFinishType {
+        let ws_stream = match connect_async(&self.url).await {
+            Ok((ws_stream, _)) => ws_stream,
+            Err(e) => {
+                ctx.report_error(
+                    "Failed to connect to websocket server".to_string(),
+                    e.to_string(),
+                )
+                .await;
+                panic!("{}", e);
+            }
+        };
+
+        let mut last_reported_error = Instant::now();
+        let mut errors = 0;
+
+        let (mut tx, mut rx) = ws_stream.split();
+
+        if let Some(msg) = &self.subscription_message {
+            if let Err(e) = tx.send(tungstenite::Message::Text(msg.clone())).await {
+                ctx.report_error(
+                    "Failed to send subscription message to websocket server".to_string(),
+                    e.to_string(),
+                )
+                .await;
+                panic!(
+                    "Failed to send subscription message to websocket server: {:?}",
+                    e
+                );
+            }
+        }
+
+        // since there's no way to partition across a websocket source, only read on the first task
+        if ctx.task_info.task_index == 0 {
+            loop {
+                select! {
+                    message = rx.next()  => {
+                        match message {
+                            Some(Ok(msg)) => {
+                                let data = match msg {
+                                    tungstenite::Message::Text(t) => {
+                                        self.serialization_mode.deserialize_str(&t).map(|t| Some(t))
+                                    },
+                                    tungstenite::Message::Binary(bs) => {
+                                        self.serialization_mode.deserialize_slice(&bs).map(|t| Some(t))
+                                    },
+                                    tungstenite::Message::Ping(d) => {
+                                        tx.send(tungstenite::Message::Pong(d)).await
+                                            .map(|_| None)
+                                            .map_err(|e| UserError::new("Failed to sned pong to websocket server", e.to_string()))
+                                    },
+                                    tungstenite::Message::Pong(_) => {
+                                        // ignore
+                                        Ok(None)
+                                    },
+                                    tungstenite::Message::Close(_) => {
+                                        ctx.report_error("Received close frame from server".to_string(), "".to_string()).await;
```

Comment:
It's not obvious to me this is an error.

Reply:
it's an error from the perspective of a long-running query; if the server tells us to go away we don't have a reliable way to continue from there